File: Code/Python_Code/07.1.1_ADC/ADC.py
#!/usr/bin/env python3
#############################################################################
# Filename    : ADC.py
# Description : Analog and Digital Conversion, ADC and DAC
# Author      : www.freenove.com
# modification: 2019/12/27
########################################################################
import smbus
import time

address = 0x48  # default address of PCF8591
bus=smbus.SMBus(1)
cmd=0x40        # command, 0100 0000

def analogRead(chn): # read ADC value,chn:0,1,2,3
    value = bus.read_byte_data(address,cmd+chn)
    return value
    
def analogWrite(value): # write DAC value
    bus.write_byte_data(address,cmd,value)  
    
def loop():
    while True:
        value = analogRead(0)   # read the ADC value of channel 0
        analogWrite(value)      # write the DAC value to control led
        voltage = value / 255.0 * 3.3  # calculate the voltage value
        print ('ADC Value : %d, Voltage : %.2f'%(value,voltage))
        time.sleep(0.01)

def destroy():
    bus.close()
    
if __name__ == '__main__':   # Program entrance
    print ('Program is starting ... ')
    try:
        loop()
    except KeyboardInterrupt: # Press ctrl-c to end the program.
        destroy()
        
    

File: Code/C_Code/07.1.1_ADC/ADC.c
/**********************************************************************
* Filename    : ADC.c
* Description : ADC and DAC
* Author      : www.freenove.com
* modification: 2019/12/27
**********************************************************************/
#include <wiringPi.h>
#include <pcf8591.h>
#include <stdio.h>

#define address 0x48  		//pcf8591 default address
#define pinbase 64			//any number above 64 (according to the wiringPi library)
#define A0 pinbase + 0
#define A1 pinbase + 1
#define A2 pinbase + 2
#define A3 pinbase + 3

int main(void){
	int value;
	float voltage;
	
	printf("Program is starting ... \n");
	
	wiringPiSetup(); //Initialize wiringPi.
	
	pcf8591Setup(pinbase,address);
	
	while(1){
		value = analogRead(A0);  //read analog value of A0 pin
		analogWrite(pinbase+0,value);
		voltage = (float)value / 255.0 * 3.3;  // Calculate voltage
		printf("ADC value : %d  ,\tVoltage : %.2fV\n",value,voltage);
		delay(100);
	}
}
